src/instance.rs

```diff
@@ -1017,14 +1017,12 @@ impl<T> Drop for Py<T> {
 impl<'a, T> FromPyObject<'a> for Py<T>
 where
     T: PyTypeInfo,
-    &'a T::AsRefTarget: FromPyObject<'a>,
-    T::AsRefTarget: 'a + AsPyPointer,
 {
-    /// Extracts `Self` from the source `PyObject`.
     fn extract(ob: &'a PyAny) -> PyResult<Self> {
-        unsafe {
-            ob.extract::<&T::AsRefTarget>()
-                .map(|val| Py::from_borrowed_ptr(ob.py(), val.as_ptr()))
+        if ob.is_instance_of::<T>() {
+            Ok(unsafe { Py::from_non_null(NonNull::new_unchecked(ob.into_ptr())) })
```

Comment:
So this can assume that `val` will always point to the same object (i.e. not a sub-allocation) as `ob`?

Reply:
Ah, that's a good observation. I think I've always assumed that should be the case, and I'd quite like this to be true. But it is breaking. So if we like this property we could document this as a breaking change?